provider/provider_yaml_test.go

```diff
@@ -459,7 +468,431 @@ func TestS3BucketObjectDeprecation(t *testing.T) {
+	}
+}
+
+// testTagsPulumiLifecycle tests the complete lifecycle of a pulumi program
+// Scenarios that this tests:
+//  1. `Up` with both provider `defaultTags`/`ignoreTags` and resource level `tags`
+//     1a. Run validations on result
+//  2. `Refresh` with no changes
+//  3. `Import` using the resource option. Ensures resource can be successfully imported
+//     3a. Allows for a hook to be run prior to import being run. e.g. Add tags remotely
+//  4. `Import` using the CLI. Ensures resources can be successfully imported
+//     4a. Allows for a hook to be run prior to import being run. e.g. Add tags remotely
+//  5. `Refresh` with no changes
+func testTagsPulumiLifecycle(t *testing.T, step tagsTestStep) {
+	t.Helper()
+	ctx := context.Background()
+
+	stepDir, err := os.MkdirTemp(os.TempDir(), step.name)
+	assert.NoError(t, err)
+	fpath := filepath.Join(stepDir, "Pulumi.yaml")
+
+	generateTagsTest(t, step, fpath, "")
+	ptest := pulumiTest(t, stepDir, opttest.TestInPlace())
+	stack := ptest.CurrentStack()
+
+	t.Log("Initial deployment...")
+	upRes, err := stack.Up(ctx)
+	assert.NoError(t, err)
+	outputs := upRes.Outputs
+	urn := outputs["urn"].Value.(string)
+	id := outputs["id"].Value.(string)
+	providerUrn := outputs["providerUrn"].Value.(string)
+	if step.postUpHook != nil {
+		step.postUpHook(t, outputs)
+	}
+
+	t.Log("refresh...")
+	_, err = stack.Refresh(ctx, optrefresh.ExpectNoChanges())
+	assert.NoError(t, err)
+
+	t.Log("delete state...")
+	execPulumi(t, ptest, stepDir, "state", "delete", urn)
+
+	// import using the import resource option
+	t.Log("up with import...")
+	if step.preImportHook != nil {
+		step.preImportHook(t, outputs)
+	}
+	generateTagsTest(t, step, fpath, id)
+	upRes, err = stack.Up(ctx, optup.Diff())
+	assert.NoError(t, err)
+	changes := *upRes.Summary.ResourceChanges
+	assert.Equal(t, 1, changes["import"])
+
+	t.Log("delete state...")
+	execPulumi(t, ptest, stepDir, "state", "delete", urn)
+
+	t.Log("import from cli...")
+	if step.preImportHook != nil {
+		step.preImportHook(t, outputs)
+	}
+	generateTagsTest(t, step, fpath, "")
+	execPulumi(t, ptest, stepDir, "import", step.typ, "res", id, "--provider", fmt.Sprintf("aws-provider=%s", providerUrn), "--yes")
+	execPulumi(t, ptest, stepDir, "state", "unprotect", urn, "--yes")
+
+	// need to run an up to fix the state. It should be a no-op
+	// re https://github.com/pulumi/pulumi-aws/issues/4204
+	upRes, err = stack.Up(ctx)
+	assert.NoError(t, err)
+	for k := range *upRes.Summary.ResourceChanges {
+		if k != "same" {
+			t.Fatal("expected no changes")
+		}
+	}
+
+	t.Log("preview with refresh...")
+	_, err = stack.Preview(ctx, optpreview.Refresh(), optpreview.ExpectNoChanges())
+	assert.NoError(t, err)
+}
+
+// generateTagsTest generates a pulumi program for the given test step
+// and writes it to the test directory
+func generateTagsTest(t *testing.T, step tagsTestStep, testPath string, importId string) {
+	template := `name: test-aws-%s
+runtime: yaml
+resources:
```

Comment:
We should test the static `get` functions as well ideally because we're changing the behavior of read

Reply:
Added!